File: Sources/FlowDictate/Models/LogEvents.swift
import Foundation

// MARK: - Log Level

enum LogLevel: String, Codable {
    case debug
    case info
    case warning
    case error
}

// MARK: - Log Category

enum LogCategory: String, Codable {
    case App
    case Hotkey
    case Audio
    case Transcription
    case Paste
    case Settings
}

// MARK: - Log Entry

struct LogEntry: Codable {
    let ts: String
    let level: LogLevel
    let appSession: String
    let dictationSession: String?
    let category: LogCategory
    let event: String
    let data: [String: AnyCodable]

    init(
        level: LogLevel,
        appSession: String,
        dictationSession: String?,
        category: LogCategory,
        event: String,
        data: [String: AnyCodable] = [:]
    ) {
        self.ts = ISO8601DateFormatter.shared.string(from: Date())
        self.level = level
        self.appSession = appSession
        self.dictationSession = dictationSession
        self.category = category
        self.event = event
        self.data = data
    }
}

// MARK: - ISO8601 Formatter

extension ISO8601DateFormatter {
    static let shared: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
}

// MARK: - Event Names

enum LogEvent {
    // App events
    enum App {
        static let started = "app_started"
        static let ready = "app_ready"
        static let terminated = "app_terminated"
    }

    // Hotkey events
    enum Hotkey {
        static let registered = "hotkey_registered"
        static let unregistered = "hotkey_unregistered"
        static let keyDown = "key_down"
        static let keyUp = "key_up"
    }

    // Audio events
    enum Audio {
        static let captureStarted = "capture_started"
        static let captureStopped = "capture_stopped"
        static let permissionGranted = "permission_granted"
        static let permissionDenied = "permission_denied"
        static let permissionRequested = "permission_requested"
        static let bufferReceived = "buffer_received"
    }

    // Transcription events
    enum Transcription {
        static let started = "transcription_started"
        static let completed = "transcription_completed"
        static let failed = "transcription_failed"
        static let modelLoading = "model_loading"
        static let modelLoaded = "model_loaded"
        static let modelFailed = "model_failed"
    }

    // Paste events
    enum Paste {
        static let attempted = "paste_attempted"
        static let succeeded = "paste_succeeded"
        static let failed = "paste_failed"
        static let permissionDenied = "permission_denied"
    }

    // Settings events
    enum Settings {
        static let changed = "settings_changed"
    }

    // Session events
    enum Session {
        static let dictationStarted = "dictation_session_started"
        static let dictationComplete = "dictation_session_complete"
        static let stateChanged = "state_changed"
    }
}

File: src-tauri/src/logging/log_events.rs
/// Log event name constants matching the Swift app's LogEvent structure.
///
/// Not every constant here is wired up yet from the Rust side — some mirror
/// events the Swift companion app emits but the Rust app_controller doesn't
/// (yet) log. Those carry their own `#[allow(dead_code)]` rather than a
/// blanket module-level one, so newly-wired constants (see `app::STARTED`,
/// `session::DICTATION_STARTED`) surface a real dead-code warning if their
/// call site is ever removed without removing the constant too.
pub mod app {
    pub const STARTED: &str = "app_started";
    #[allow(dead_code)]
    pub const READY: &str = "app_ready";
    #[allow(dead_code)]
    pub const TERMINATED: &str = "app_terminated";
}

#[allow(dead_code)]
pub mod hotkey {
    pub const REGISTERED: &str = "hotkey_registered";
    pub const UNREGISTERED: &str = "hotkey_unregistered";
    pub const KEY_DOWN: &str = "key_down";
    pub const KEY_UP: &str = "key_up";
}

#[allow(dead_code)]
pub mod audio {
    pub const CAPTURE_STARTED: &str = "capture_started";
    pub const CAPTURE_STOPPED: &str = "capture_stopped";
    pub const PERMISSION_GRANTED: &str = "permission_granted";
    pub const PERMISSION_DENIED: &str = "permission_denied";
}

#[allow(dead_code)]
pub mod transcription {
    pub const STARTED: &str = "transcription_started";
    pub const COMPLETED: &str = "transcription_completed";
    pub const FAILED: &str = "transcription_failed";
    pub const MODEL_LOADING: &str = "model_loading";
    pub const MODEL_LOADED: &str = "model_loaded";
}

#[allow(dead_code)]
pub mod paste {
    pub const ATTEMPTED: &str = "paste_attempted";
    pub const SUCCEEDED: &str = "paste_succeeded";
    pub const FAILED: &str = "paste_failed";
}

pub mod session {
    pub const DICTATION_STARTED: &str = "dictation_session_started";
    #[allow(dead_code)]
    pub const DICTATION_COMPLETE: &str = "dictation_session_complete";
    #[allow(dead_code)]
    pub const STATE_CHANGED: &str = "state_changed";
}
